chips/stm32f446re/src/chip.rs

```diff
@@ -0,0 +1,109 @@
+        }
+    }
+}
+
+impl Chip for Stm32f446re {
+    type MPU = cortexm4::mpu::MPU;
+    type UserspaceKernelBoundary = cortexm4::syscall::SysCall;
+    type SysTick = cortexm4::systick::SysTick;
+
+    fn dispatch_interrupt_events_and_deferred_tasks(&self) {
+        deferred_call_tasks::dispatch_deferred_call_tasks();
+        interrupt_events::dispatch_interrupt_events();
+    }
+
+    fn has_undispatched_interrupt_events_or_deferred_tasks(&self) -> bool {
+        interrupt_events::has_undispatched_interrupt_events() || deferred_call::has_tasks()
+    }
+
+    fn mpu(&self) -> &cortexm4::mpu::MPU {
+        &self.mpu
+    }
+
+    fn systick(&self) -> &cortexm4::systick::SysTick {
+        &self.systick
+    }
+
+    fn userspace_kernel_boundary(&self) -> &cortexm4::syscall::SysCall {
+        &self.userspace_kernel_boundary
+    }
+
+    fn sleep(&self) {
+        // From power controller/manager to see if "deep sleep" is possible. If
+        // so, enable it here.
+        //
+        // The only place this *should* be called from is from `kernel_loop`.
+        // Before this is called, we need to *ensure* that we are in a critical
+        // section. This can be done using `atomic`.
+        //
+        // The main reason for sleeping inside a critical section is that we
+        // want to be deterministically woken up right after the `wfi`.
+        //
+        // The high level logic should be as follows -
+        //
+        // 1. Enter critical section (handled by `atomic`)
+        //
+        // 2. Ensure that we do not have any user space work to do. If we do,
+        //    exit critical section and re-enter the loop
+        //
+        //    This part is done in `kernel_loop`.
+        //
+        // 3. Ensure that INTERRUPT_EVENTS is 0. This signifies two things -
+        //
+        //    a. All interrupts bottom halves that are waiting to be dispatched
+        //       have been dispatched.
+        //
+        //    b. If an interrupt occurs at this stage or later, then we will
+        //       wake up right after `wfi` or `wfi` will become a `nop`
+        //
+        //    If INTERRUPT_EVENTS is not 0, then we have undispatched
+        //    interrupts, so exit the critical section an get back to kernel
+        //    loop.
+        //
+        //    This part is done in `kernel_loop`
+        //
+        // 4. Do any vendor specific sleep configuration settings. At a minimum
+        //    this involves clearing SCB' deepsleep bit to prevent us from
+        //    accidentally going to deep sleep. In normal sleep, its only the
+        //    clock to the CPU that is gated. Clock is still available Systick
+        //    and NVIC peripherals is still available, which is what we want.
+        //
+        // 5. Do a DSB (If required, not needed in case of STM32F446RE)
+        //
+        // 6. Do a WFI (Go to sleep)
+        //
+        // 7. Exit critical section (handled by `atomic`)
```

Comment:
This is a nice overview, but I'm wary of having so much kernel-specific stuff (e.g. interrupt bottom half queue) spelled out in a chip driver -- if it ever changes this doc is likely to bit rot :/

Consider moving some of this comment into the kernel loop itself, or the Tock documentation, and try to pare this down to just the STM32F-specific bits?